lib/mneme/assertion.ex

```diff
@@ -96,6 +133,113 @@ defmodule Mneme.Assertion do
     end
   end
 
+  # Prevents warnings about unused aliases when their only usage is in an auto-assertion
+  defp silence_used_aliases(macro_ast, context_aliases) do
+    macro_ast
+    |> extract_used_aliases(context_aliases)
+    |> Enum.map(&quoted_dummy_assign/1)
+  end
+
+  defp collect_vars_from_pattern({:when, _, [left, right]}) do
+    pattern = collect_vars_from_pattern(left)
+
+    vars =
+      for {name, _, context} = var <- collect_vars_from_pattern(right),
+          has_var?(pattern, name, context),
+          do: var
+
+    pattern ++ vars
+  end
+
+  defp collect_vars_from_pattern(expr) do
+    expr
+    |> Macro.prewalk([], fn
+      {:"::", _, [left, right]}, acc ->
+        {[left], collect_vars_from_binary(right, acc)}
+
+      {skip, _, [_]}, acc when skip in [:^, :@, :quote] ->
+        {:ok, acc}
+
+      {skip, _, [_, _]}, acc when skip in [:quote] ->
+        {:ok, acc}
+
+      {:_, _, context}, acc when is_atom(context) ->
+        {:ok, acc}
+
+      {name, meta, context}, acc when is_atom(name) and is_atom(context) ->
+        {:ok, [{name, meta, context} | acc]}
+
+      node, acc ->
+        {node, acc}
+    end)
+    |> elem(1)
+  end
+
+  defp collect_vars_from_binary(right, original_acc) do
+    right
+    |> Macro.prewalk(original_acc, fn
+      {mode, _, [{name, meta, context}]}, acc
+      when is_atom(mode) and is_atom(name) and is_atom(context) ->
+        if has_var?(original_acc, name, context) do
+          {:ok, [{name, meta, context} | acc]}
+        else
+          {:ok, acc}
+        end
+
+      node, acc ->
+        {node, acc}
+    end)
+    |> elem(1)
+  end
+
+  defp has_var?(pattern, name, context), do: Enum.any?(pattern, &match?({^name, _, ^context}, &1))
+
+  defp mark_as_generated(vars) do
+    for {name, meta, context} <- vars, do: {name, [generated: true] ++ meta, context}
+  end
+
+  @doc false
```

Comment:
_NB_ `__expand_pattern__/2` has been stolen as is from `ExUnit.Assertions`, which is also public there, so we might instead rely on the direct call to `ExUnit.Assertions.__expand_pattern__/2` above.

I decided to copy-paste code here because it might require some customization for the fixed-in-the-future assertions beyond simple `_ <- _`.